shared/src/commonMain/kotlin/org/hyperskill/app/search/presentation/SearchReducer.kt

```diff
@@ -0,0 +1,130 @@
+                handleSearchResultsItemClickedMessage(state, message)
+            }
+            Message.ViewedEventMessage -> {
+                state to setOf(InternalAction.LogAnalyticEvent(SearchViewedHyperskillAnalyticEvent))
+            }
+        } ?: (state to emptySet())
+
+    private fun handleQueryChangedMessage(
+        state: State,
+        message: Message.QueryChanged
+    ): SearchReducerResult? {
+        if (state.searchResultsState == SearchFeature.SearchResultsState.Loading) {
+            return null
+        }
+
+        val newSearchResultsState = when (state.searchResultsState) {
+            SearchFeature.SearchResultsState.Editing -> SearchFeature.SearchResultsState.Editing
+            SearchFeature.SearchResultsState.Loading -> SearchFeature.SearchResultsState.Loading
+            SearchFeature.SearchResultsState.Error -> SearchFeature.SearchResultsState.Editing
+            is SearchFeature.SearchResultsState.Content -> SearchFeature.SearchResultsState.Editing
+        }
+
+        return state.copy(
+            query = message.query,
+            searchResultsState = newSearchResultsState
+        ) to emptySet()
+    }
+
+    private fun handleSearchClickedMessage(state: State): SearchReducerResult {
+        val analyticActions = setOf(
+            InternalAction.LogAnalyticEvent(SearchClickedSearchHyperskillAnalyticEvent(state.query))
+        )
+
+        if (state.query.isBlank()) {
+            return state to analyticActions
+        }
+
+        return when (state.searchResultsState) {
+            SearchFeature.SearchResultsState.Editing,
+            SearchFeature.SearchResultsState.Error ->
+                state.copy(
+                    searchResultsState = SearchFeature.SearchResultsState.Loading
+                ) to (analyticActions + InternalAction.PerformSearch(state.query))
+            SearchFeature.SearchResultsState.Loading,
+            is SearchFeature.SearchResultsState.Content ->
+                state to analyticActions
+        }
+    }
+
+    private fun handleSearchResultsItemClickedMessage(
+        state: State,
+        message: Message.SearchResultsItemClicked
+    ): SearchReducerResult? {
+        if (state.searchResultsState is SearchFeature.SearchResultsState.Content) {
+            val targetTopic = state.searchResultsState.topics.firstOrNull { it.id == message.id } ?: return null
+            return state to buildSet {
+                if (targetTopic.theoryId != null) {
+                    add(
+                        Action.ViewAction.OpenStepScreen(
+                            StepRoute.Learn.TheoryOpenedFromSearch(targetTopic.theoryId)
+                        )
+                    )
+                }
+                add(
+                    InternalAction.LogAnalyticEvent(
+                        SearchClickedItemHyperskillAnalyticEvent(query = state.query, topicId = targetTopic.id)
+                    )
+                )
+            }
```

Comment:
I suggest to add error viewAction, in case theoryId == null, to let the user know that something went wrong.